com.unity.render-pipelines.universal/Runtime/RendererFeatures/ScreenSpaceShadows.cs

```diff
@@ -0,0 +1,230 @@
+using System;
+using UnityEngine.Experimental.Rendering;
+
+namespace UnityEngine.Rendering.Universal
+{
+    [Serializable]
+    internal class ScreenSpaceShadowsSettings
+    {
+
+    }
+
+    [DisallowMultipleRendererFeature]
+    internal class ScreenSpaceShadows : ScriptableRendererFeature
+    {
+        // Serialized Fields
+        [SerializeField, HideInInspector] private Shader m_Shader = null;
+        [SerializeField] private ScreenSpaceShadowsSettings m_Settings = new ScreenSpaceShadowsSettings();
+
+        // Private Fields
+        private Material m_Material;
+        private ScreenSpaceShadowsPass m_SSShadowsPass = null;
+        private EnableKeywordPass m_EnableKeywordPass = null;
+        private DisableKeywordPass m_DisableKeywordPass = null;
+
+        // Constants
+        private const string k_ShaderName = "Hidden/Universal Render Pipeline/ScreenSpaceShadows";
+
+        /// <inheritdoc/>
+        public override void Create()
+        {
+            if (m_SSShadowsPass == null)
+                m_SSShadowsPass = new ScreenSpaceShadowsPass();
+            if (m_EnableKeywordPass == null)
+                m_EnableKeywordPass = new EnableKeywordPass();
+            if (m_DisableKeywordPass == null)
+                m_DisableKeywordPass = new DisableKeywordPass();
+
+            LoadMaterial();
+            m_SSShadowsPass.profilerTag = name;
+            m_SSShadowsPass.renderPassEvent = RenderPassEvent.AfterRenderingPrepasses;
+
+            m_EnableKeywordPass.renderPassEvent = RenderPassEvent.BeforeRenderingOpaques;
+            m_DisableKeywordPass.renderPassEvent = RenderPassEvent.AfterRenderingOpaques;
+        }
+
+        /// <inheritdoc/>
+        public override void AddRenderPasses(ScriptableRenderer renderer, ref RenderingData renderingData)
+        {
+            if (m_Material == null)
+            {
+                Debug.LogErrorFormat(
+                    "{0}.AddRenderPasses(): Missing material. {1} render pass will not be added. Check for missing reference in the renderer resources.",
+                    GetType().Name, m_SSShadowsPass.profilerTag);
+                return;
+            }
+
+            bool shouldAdd = m_SSShadowsPass.Setup(m_Settings) && renderingData.shadowData.supportsMainLightShadows;
+            if (shouldAdd)
+            {
+                renderer.EnqueuePass(m_SSShadowsPass);
+                renderer.EnqueuePass(m_EnableKeywordPass);
+                renderer.EnqueuePass(m_DisableKeywordPass);
+            }
+        }
+
+        /// <inheritdoc/>
+        protected override void Dispose(bool disposing)
+        {
+            CoreUtils.Destroy(m_Material);
+        }
+
+        private bool LoadMaterial()
```

Comment:
Suggestion: today we reference the screen space shadow shader in the renderer asset. (Shaders struct). Since now we have a feature for it we could reference the shader in the feature. This way, we only include the shader in the build if the feature is added to any renderer as opposed to today that we include it even when we don't use it.

Reply:
```csharp
            [Reload("Shaders/Utils/ScreenSpaceShadows.shader")]
            public Shader screenSpaceShadowPS;
```

Does you mean the shader in renderer is needed to be removed?
I followed a way in screen space ambient occlusion

Reply:
hm, if the API is public we can't remove it (should make obsolete in that case), but can remove the reference and the [Reload] attribute.

Reply:
Likewise you can add the same construct to your ScreenSpaceShadow shader so you make sure that you include the shader in the build if the feature is reference in any renderer.

Reply:
Add obsolete attribute and remove reload and reference, also remove material of screen space shadow in forward and deferred renderer